File: solutions/0350-intersection-of-two-arrays-ii/hash-table.js
// time: O(n + m)
// space: O(min(n,m))

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
const intersect = (nums1, nums2) => {
    const freq = {};
    const result = [];

    for (const num of nums1) freq[num] = (freq[num] || 0) + 1;
    for (const num of nums2) {
        if (freq[num]) {
            result.push(num);
            freq[num]--;
        }
    }

    return result;
};

File: solutions/0350-intersection-of-two-arrays-ii/hash-table.cpp
// time: O(n + m)
// space: O(min(n,m))

#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
  public:
    vector<int> intersect(vector<int> &nums1, vector<int> &nums2) {
        unordered_map<int, int> freq;
        vector<int> result;

        for (const int num : nums1) freq[num]++;
        for (const int num : nums2) {
            if (freq[num]) {
                result.push_back(num);
                freq[num]--;
            }
        }

        return result;
    }
};
